File: solver/maze.py
"""Maze solver algorithms (BFS, DFS, A*)."""

import random
import time
from collections import deque
from typing import List, Dict, Tuple, Set
from dataclasses import dataclass

from .common import MinHeap, SolveStats

Cell = int  # 0 = path, 1 = wall
Maze = List[List[Cell]]


@dataclass
class Pos:
    """Position in maze."""

    r: int
    c: int

    def __hash__(self):
        return hash((self.r, self.c))

    def __eq__(self, other):
        if isinstance(other, Pos):
            return self.r == other.r and self.c == other.c
        return False

    def __lt__(self, other):
        return (self.r, self.c) < (other.r, other.c)


@dataclass
class MazeData:
    """Maze data structure."""

    grid: Maze
    start: Pos
    end: Pos
    rows: int
    cols: int


@dataclass
class MazeStep:
    """Single step in maze exploration."""

    explored: List[Pos]
    path: List[Pos]


@dataclass
class MazeSolveResult:
    """Result of maze solving."""

    order: List[Pos]
    path: List[Pos]
    stats: SolveStats


def pos_key(pos: Pos) -> str:
    """Convert position to key."""
    return f"{pos.r},{pos.c}"


def get_neighbors(grid: Maze, pos: Pos) -> List[Pos]:
    """Get valid neighbors of a position."""
    neighbors = []
    rows, cols = len(grid), len(grid[0])

    for dr, dc in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
        nr, nc = pos.r + dr, pos.c + dc
        if 0 <= nr < rows and 0 <= nc < cols and grid[nr][nc] == 0:
            neighbors.append(Pos(nr, nc))

    return neighbors


def generate_maze(difficulty: str) -> MazeData:
    """Generate maze using DFS recursive backtracker."""
    size_map = {"easy": 11, "medium": 19, "hard": 27}
    size = size_map.get(difficulty, 19)
    rows, cols = size, size

    # Initialize grid with all walls
    grid: Maze = [[1 for _ in range(cols)] for _ in range(rows)]

    # DFS backtracker
    stack = [Pos(1, 1)]
    grid[1][1] = 0

    while stack:
        current = stack[-1]
        # All four directions with random order
        directions = [(-2, 0), (2, 0), (0, -2), (0, 2)]
        random.shuffle(directions)

        carved = False
        for dr, dc in directions:
            nr, nc = current.r + dr, current.c + dc

            if (
                0 < nr < rows - 1
                and 0 < nc < cols - 1
                and grid[nr][nc] == 1
            ):
                # Carve path between current and next
                grid[current.r + dr // 2][current.c + dc // 2] = 0
                grid[nr][nc] = 0
                stack.append(Pos(nr, nc))
                carved = True
                break

        if not carved:
            stack.pop()

    return MazeData(
        grid=grid,
        start=Pos(1, 1),
        end=Pos(rows - 2, cols - 2),
        rows=rows,
        cols=cols,
    )


def reconstruct_path(
    start: Pos, end: Pos, parent: Dict[str, str], found: bool
) -> List[Pos]:
    """Reconstruct path from parent map."""
    if not found:
        return []

    path = []
    current_key = pos_key(end)

    while current_key:
        r, c = map(int, current_key.split(","))
        path.append(Pos(r, c))

        if current_key == pos_key(start):
            break

        current_key = parent.get(current_key)

    return list(reversed(path))


def solve_bfs(maze: MazeData) -> MazeSolveResult:
    """Solve maze using BFS."""
    start_time = time.time()
    parent: Dict[str, str] = {}
    visited: Set[str] = {pos_key(maze.start)}
    order: List[Pos] = []
    queue = deque([maze.start])
    found = False

    while queue:
        current = queue.popleft()
        order.append(current)

        if current.r == maze.end.r and current.c == maze.end.c:
            found = True
            break

        for neighbor in get_neighbors(maze.grid, current):
            key = pos_key(neighbor)
            if key not in visited:
                visited.add(key)
                parent[key] = pos_key(current)
                queue.append(neighbor)

    path = reconstruct_path(maze.start, maze.end, parent, found)
    elapsed = (time.time() - start_time) * 1000

    stats = SolveStats(
        algorithm="BFS",
        steps=max(0, len(path) - 1),
        nodes_explored=len(order),
        time_ms=elapsed,
        found=found,
    )

    return MazeSolveResult(order=order, path=path, stats=stats)


def solve_dfs(maze: MazeData) -> MazeSolveResult:
    """Solve maze using DFS."""
    start_time = time.time()
    parent: Dict[str, str] = {}
    visited: Set[str] = {pos_key(maze.start)}
    order: List[Pos] = []
    stack = [maze.start]
    found = False

    while stack:
        current = stack.pop()
        order.append(current)

        if current.r == maze.end.r and current.c == maze.end.c:
            found = True
            break

        for neighbor in get_neighbors(maze.grid, current):
            key = pos_key(neighbor)
            if key not in visited:
                visited.add(key)
                parent[key] = pos_key(current)
                stack.append(neighbor)

    path = reconstruct_path(maze.start, maze.end, parent, found)
    elapsed = (time.time() - start_time) * 1000

    stats = SolveStats(
        algorithm="DFS",
        steps=max(0, len(path) - 1),
        nodes_explored=len(order),
        time_ms=elapsed,
        found=found,
    )

    return MazeSolveResult(order=order, path=path, stats=stats)


def solve_astar(maze: MazeData) -> MazeSolveResult:
    """Solve maze using A*."""
    start_time = time.time()

    def heuristic(pos: Pos) -> int:
        """Manhattan distance heuristic."""
        return abs(pos.r - maze.end.r) + abs(pos.c - maze.end.c)

    heap = MinHeap()
    g_score: Dict[str, int] = {pos_key(maze.start): 0}
    parent: Dict[str, str] = {}
    closed: Set[str] = set()
    order: List[Pos] = []

    start_h = heuristic(maze.start)
    heap.push(start_h, maze.start)
    found = False

    while not heap.empty():
        current = heap.pop()
        current_key = pos_key(current)

        if current_key in closed:
            continue

        closed.add(current_key)
        order.append(current)

        if current.r == maze.end.r and current.c == maze.end.c:
            found = True
            break

        current_g = g_score[current_key]

        for neighbor in get_neighbors(maze.grid, current):
            key = pos_key(neighbor)
            tentative_g = current_g + 1

            if tentative_g < g_score.get(key, float("inf")):
                g_score[key] = tentative_g
                parent[key] = current_key
                f_score = tentative_g + heuristic(neighbor)
                heap.push(f_score, neighbor)

    path = reconstruct_path(maze.start, maze.end, parent, found)
    elapsed = (time.time() - start_time) * 1000

    stats = SolveStats(
        algorithm="A*",
        steps=max(0, len(path) - 1),
        nodes_explored=len(order),
        time_ms=elapsed,
        found=found,
    )

    return MazeSolveResult(order=order, path=path, stats=stats)

File: src/backend/ai/maze.ts
import { MinHeap, SolveStats } from "./common";

export type Cell = 0 | 1; // 0 path, 1 wall
export type Maze = Cell[][];
export type Pos = { r: number; c: number };

export interface MazeData { grid: Maze; start: Pos; end: Pos; rows: number; cols: number }

// DFS recursive backtracker generation (perfect maze, always solvable)
export const generateMaze = (difficulty: "easy" | "medium" | "hard"): MazeData => {
  const size = difficulty === "easy" ? 11 : difficulty === "medium" ? 19 : 27; // odd sizes
  const rows = size, cols = size;
  const g: Maze = Array.from({ length: rows }, () => Array(cols).fill(1) as Cell[]);
  const stack: Pos[] = [{ r: 1, c: 1 }];
  g[1][1] = 0;
  while (stack.length) {
    const cur = stack[stack.length - 1];
    const dirs = [[-2,0],[2,0],[0,-2],[0,2]].sort(() => Math.random() - 0.5);
    let carved = false;
    for (const [dr, dc] of dirs) {
      const nr = cur.r + dr, nc = cur.c + dc;
      if (nr > 0 && nr < rows - 1 && nc > 0 && nc < cols - 1 && g[nr][nc] === 1) {
        g[cur.r + dr / 2][cur.c + dc / 2] = 0;
        g[nr][nc] = 0;
        stack.push({ r: nr, c: nc }); carved = true; break;
      }
    }
    if (!carved) stack.pop();
  }
  return { grid: g, start: { r: 1, c: 1 }, end: { r: rows - 2, c: cols - 2 }, rows, cols };
};

const key = (p: Pos) => `${p.r},${p.c}`;
const neighbors = (g: Maze, p: Pos): Pos[] => {
  const out: Pos[] = [];
  for (const [dr, dc] of [[-1,0],[1,0],[0,-1],[0,1]]) {
    const nr = p.r + dr, nc = p.c + dc;
    if (nr >= 0 && nr < g.length && nc >= 0 && nc < g[0].length && g[nr][nc] === 0) out.push({ r: nr, c: nc });
  }
  return out;
};

export interface MazeStep { explored: Pos[]; path: Pos[] }

export interface MazeSolveResult { order: Pos[]; path: Pos[]; stats: SolveStats }

export const solveBFS = (m: MazeData): MazeSolveResult => {
  const t0 = performance.now();
  const parent = new Map<string, string>();
  const visited = new Set<string>([key(m.start)]);
  const order: Pos[] = [];
  const queue: Pos[] = [m.start]; let head = 0; let found = false;
  while (head < queue.length) {
    const cur = queue[head++]; order.push(cur);
    if (cur.r === m.end.r && cur.c === m.end.c) { found = true; break; }
    for (const n of neighbors(m.grid, cur)) {
      const k = key(n); if (!visited.has(k)) { visited.add(k); parent.set(k, key(cur)); queue.push(n); }
    }
  }
  const path = reconstructPath(m.start, m.end, parent, found);
  return { order, path, stats: { algorithm: "BFS", steps: Math.max(0, path.length - 1), nodesExplored: order.length, timeMs: +(performance.now() - t0).toFixed(2), found } };
};

export const solveDFS = (m: MazeData): MazeSolveResult => {
  const t0 = performance.now();
  const parent = new Map<string, string>();
  const visited = new Set<string>([key(m.start)]);
  const order: Pos[] = [];
  const stack: Pos[] = [m.start]; let found = false;
  while (stack.length) {
    const cur = stack.pop()!; order.push(cur);
    if (cur.r === m.end.r && cur.c === m.end.c) { found = true; break; }
    for (const n of neighbors(m.grid, cur)) {
      const k = key(n); if (!visited.has(k)) { visited.add(k); parent.set(k, key(cur)); stack.push(n); }
    }
  }
  const path = reconstructPath(m.start, m.end, parent, found);
  return { order, path, stats: { algorithm: "DFS", steps: Math.max(0, path.length - 1), nodesExplored: order.length, timeMs: +(performance.now() - t0).toFixed(2), found } };
};

export const solveAStar = (m: MazeData): MazeSolveResult => {
  const t0 = performance.now();
  const h = (p: Pos) => Math.abs(p.r - m.end.r) + Math.abs(p.c - m.end.c);
  const heap = new MinHeap<Pos>();
  const g = new Map<string, number>([[key(m.start), 0]]);
  const parent = new Map<string, string>();
  const closed = new Set<string>();
  const order: Pos[] = [];
  heap.push(h(m.start), m.start);
  let found = false;
  while (heap.size()) {
    const cur = heap.pop()!; const ck = key(cur);
    if (closed.has(ck)) continue; closed.add(ck); order.push(cur);
    if (cur.r === m.end.r && cur.c === m.end.c) { found = true; break; }
    for (const n of neighbors(m.grid, cur)) {
      const k = key(n); const t = (g.get(ck) ?? 0) + 1;
      if (t < (g.get(k) ?? Infinity)) { g.set(k, t); parent.set(k, ck); heap.push(t + h(n), n); }
    }
  }
  const path = reconstructPath(m.start, m.end, parent, found);
  return { order, path, stats: { algorithm: "A*", steps: Math.max(0, path.length - 1), nodesExplored: order.length, timeMs: +(performance.now() - t0).toFixed(2), found } };
};

const reconstructPath = (start: Pos, end: Pos, parent: Map<string, string>, found: boolean): Pos[] => {
  if (!found) return [];
  const path: Pos[] = [];
  let k: string | undefined = key(end);
  while (k) {
    const [r, c] = k.split(",").map(Number); path.push({ r, c });
    if (k === key(start)) break;
    k = parent.get(k);
  }
  return path.reverse();
};
